lib/babe/babe.go

```diff
@@ -154,65 +158,82 @@ func NewService(cfg *ServiceConfig) (*Service, error) {
-	if cfg.SlotDuration > 0 {
+	if cfg.SlotDuration > 0 && cfg.IsDev { // TODO: remove this, needs to be set via runtime
 		b.slotDuration, err = time.ParseDuration(fmt.Sprintf("%dms", cfg.SlotDuration))
+	} else if cfg.SlotDuration > 0 && !cfg.IsDev {
+		err = errors.New("slot duration modified in config for non-dev chain")
 	} else {
-		b.slotDuration, err = time.ParseDuration(fmt.Sprintf("%dms", genCfg.SlotDuration))
+		b.slotDuration, err = b.epochState.GetSlotDuration()
 	}
 	if err != nil {
 		return err
 	}
 
-	if cfg.AuthData == nil {
-		b.epochData.authorities, err = types.BABEAuthorityRawToAuthority(genCfg.GenesisAuthorities)
-		if err != nil {
-			return err
-		}
+	if cfg.EpochLength != 0 && cfg.IsDev { // TODO: remove this, needs to be set via runtime
+		b.epochLength = cfg.EpochLength
+	} else if cfg.EpochLength > 0 && !cfg.IsDev {
+		err = errors.New("epoch length modified in config for non-dev chain")
 	} else {
-		b.epochData.authorities = cfg.AuthData
+		b.epochLength, err = b.epochState.GetEpochLength()
 	}
-
-	if cfg.Authority {
-		b.epochData.authorityIndex, err = b.getAuthorityIndex(b.epochData.authorities)
-		if err != nil {
-			return err
-		}
+	if err != nil {
+		return err
 	}
 
-	if cfg.ThresholdDenominator == 0 {
-		b.epochData.threshold, err = CalculateThreshold(genCfg.C1, genCfg.C2, len(b.epochData.authorities))
+	if cfg.AuthData != nil && cfg.IsDev { // TODO: remove this, needs to be set via runtime
+		b.epochData.authorities = cfg.AuthData
+	} else if cfg.AuthData != nil && !cfg.IsDev {
+		return errors.New("authority data modified in config for non-dev chain")
 	} else {
-		b.epochData.threshold, err = CalculateThreshold(cfg.ThresholdNumerator, cfg.ThresholdDenominator, len(b.epochData.authorities))
+		b.epochData.authorities = epochData.Authorities
 	}
 
+	if cfg.ThresholdDenominator != 0 && cfg.IsDev { // TODO: remove this, needs to be set via runtime
+		b.epochData.threshold, err = CalculateThreshold(cfg.ThresholdNumerator, cfg.ThresholdDenominator, len(b.epochData.authorities))
+	} else if cfg.ThresholdDenominator != 0 && !cfg.IsDev {
+		err = errors.New("threshold modified in config for non-dev chain")
+	} else {
+		b.epochData.threshold, err = CalculateThreshold(configData.C1, configData.C2, len(b.epochData.authorities))
+	}
 	if err != nil {
 		return err
 	}
 
-	if cfg.EpochLength > 0 {
-		b.epochLength = cfg.EpochLength
-	} else {
-		b.epochLength = genCfg.EpochLength
+	if !cfg.Authority {
+		return nil
 	}
 
-	return nil
+	b.epochData.authorityIndex, err = b.getAuthorityIndex(b.epochData.authorities)
+	return err
 }
 
 // Start starts BABE block authoring
 func (b *Service) Start() error {
-	epoch, err := b.epochState.GetCurrentEpoch()
-	if err != nil {
-		logger.Error("failed to get current epoch", "error", err)
-		return err
+	if !b.authority {
+		return nil
 	}
 
-	go b.initiate(epoch)
+	// wait a bit to check if we need to sync before initiating
+	time.Sleep(initialWaitTime)
```

Comment:
```suggestion
	<- time.NewTimer(initialWaitTime).C
```

Reply:
is this more efficient?

Reply:
updated btw, just curious about the difference

Reply:
I think the `Start` function form `babe` package is not resource-intensive, `time.NewTimer` or `time.After` will fit well than `time.Sleep`. obs: in the case of `time.NewTimer`, would be a good practice calls `defer t.Stop()`

Reply:
this will always block until the timer fires so there isn't a point to calling `Stop` here afaik